Avoid some unnecessary deactivations due to routes mismatches

ihuguet opened the pull request:
When comparing the routes with those already added to NM, sometimes
nmstate considered different routes that were actually equal.
- route1 with table=None vs route2 with table=Some(254)
- route1 with next-hop=None vs route2 with next-hop=Some(0.0.0.0)
- route1 is ipv6 with metric=0, route2 is ipv6 with metric=1024 (see explanation below).

The consequence was that unnecessary deactivate/reactivate were done,
instead of reapply. This was because is_route_removed was returning true
with some of the situations described above. This caused that
gen_nm_conn_need_to_deactivate_first also returned true and the
connection was deactivated.

For next-hop, fix it by normalizing to None at creation time, since 0.0.0.0 and ::
are not used by NM nor by the kernel, they just leave it undefined.

For table, fix it by normalizing to Some(254) because we need to always set
the table for a different reason: if the user doesn't define the table in the
desired state we are supposed to save it to the 254 main table. If we left
table empty, it might happen that a different table is used if auto-route-table-id
is also used.

For IPv6 routes metric is also included. When setting metric=0 on a kernel's IPv6 route,
kernel coherces it to 1024. This make that, when reading back the route from kernel
it doesn't match with the expected value of 0. The fix normalizes 0->1024 for IPv6
routes so the comparison doesn't fail.

Additionally, when comparing NmIpRoutes, this PR add code to handle it correctly
for every possible NMConnection. If the route's table is undefined, it falls back to
ipv4/6.route-table, as this is what NM does. Note that for routes defined by
nmstate this won't happen because, since this PR, we always define the table; but
it will handle better the case of already existing NMConnections. The same for the
metric with route-metric.

Note that this only affects NM specific code, not the RouteEntry
entities from the top level nmstate.

Resolves: https://issues.redhat.com/browse/RHEL-65004

cathay4t commented on rust/src/lib/nm/nm_dbus/connection/route.rs:
```diff
@@ -26,18 +30,32 @@ pub struct NmIpRoute {
+        let dest = _from_map!(v, "dest", String::try_from)?;
+        let default_next_hop = dest.as_deref().and_then(|dest| {
+            match dest.parse::<std::net::IpAddr>() {
+                Ok(std::net::IpAddr::V4(_)) => Some(IPV4_EMPTY_NEXT_HOP_ADDR),
```
```rust
Ok(std::net::IpAddr::V4(i)) => Some(i.to_string())
```

ihuguet replied:
If I do that, every time an `NmIpRoute` is created we will have a String allocation, but in most cases it won't be used. As I have done, here I return an `Option<&str>` and only call `to_string()` if it's needed, which is not often because almost all routes have next-hop values.

ihuguet replied:
Ah, no, I didn't understood your comment. With your change I would use the value of `dest` as "empty next-hop" value, but I have to use IPV6_EMPTY_NEXT_HOP_ADDR / IPV6_EMPTY_NEXT_HOP_ADDR instead.

cathay4t commented on rust/src/lib/nm/nm_dbus/connection/route.rs:
```diff
@@ -26,18 +30,32 @@ pub struct NmIpRoute {
     fn try_from(mut v: DbusDictionary) -> Result<Self, Self::Error> {
+        let dest = _from_map!(v, "dest", String::try_from)?;
+        let default_next_hop = dest.as_deref().and_then(|dest| {
+            match dest.parse::<std::net::IpAddr>() {
```
Base on your code it seems you are assuming `dest` in NM is always 0.0.0.0 or ::. Is this correct?

cathay4t replied:
And `nm_dbus` should not have this modification code, it should be simple dbus wrapper. This polish/modification should be done at `nm` level.

ihuguet replied:
No. Internally, nmstate represents "empty next-hop" as `Some("0.0.0.0")` for IPv4 and `Some("::")` for IPv6. So what I do here is detecting whether it's IPv4 or 6 depending on `dest`, so I can know what the value for "empty next-hop" should be.

Nmstate is doing this because `None` means "ignore next-hop value for matching", not "empty next-hop".

Where can I do the modification at `nm` level? This is where the `NmIpRoute`'s are created, so what I am doing is to create them always with an unified criteria for "default" or "empty" values.

ihuguet replied:
I don't see how to do this at `nm` level. What it's failing is the comparison between `NmIpRoutes`. This is only defined at `nm_dbus` level and it's almost not used outside of `nm_dbus`. Only a few minor usages from `nm/query_apply` and `nm/settings`.

ihuguet replied:
Fixed all at `nm` level. I've left a small normalization in `nm_dbus`, but it's a condition that should never happen and I've added a warning message so we can notice it it starts happening at some point, so it will be easy to debug.

cathay4t commented on tests/integration/route_test.py:
```diff
@@ -143,6 +156,75 @@ def test_add_static_route_without_next_hop_address(eth1_up):
+    # empty next-hop: '0.0.0.0' or '::'.
+    # If nmstate wouldn't consider them equal it would do deactivate/reactivate.
+    # Ensure that we only reapply with ip_monitor_assert_stable_link_up.
+    _nmcli_con_modify_routes_reapply("eth1", IPV4_TEST_NET1, IPV6_TEST_NET1)
```
Can you explain why use `nmcli modify` instead apply another nmstate YAML?

I think the original issue report does not use nmcli.

ihuguet replied:
Because if I use nmstate YAML it is dependant to the internal implementation of nmstate how to create the NM connection. I have seen that in some cases it is adding table=254, for example, while I want to test the case that the NM profile exists without `table`. So what I do here is to force certain combinations of initial NM's state of NM and nmstate desired states.

cathay4t replied:
Please consider use nmstate to reproduce the issue as the reporter did.
Otherwise, move the test case to `tests/integration/nm` folder as nm specific tests which will be dropped when we move away from NM1.

So please consider reproduce this issue use nmstate only.

ihuguet replied:
The bug is caused by NM specific code, it cannot be reliably reproduced with nmstate only. Further modifications to nmstate could make that the integration test always pass, even if the bug were reintroduced.
Moved to /nm subfolder.

ihuguet replied:
Changed the test about the next-hop to use nmstate apply only. The test about the table was better using nmcli, I think, and also with the newest changes it cannot be reproduced with nmstate only any more, so it has to remain as NM specific and use nmcli.

ihuguet commented:
I have changed to a different approach: instead of normalizing to "Some(default_value)", normalize to "None" if the value is the default. This was what we were doing for "weight" and it makes sense to do it also for "table" and "next-hop".

ffmancera reviewed (approved):
To me the patch looks good, I did a small test with the reproducer I got and it works.

I prefer this new solution, it makes sense to normalize the defaults to None as weigh was doing.

liangwen12year commented on rust/src/lib/nm/query_apply/route.rs:
```diff
@@ -37,14 +35,7 @@ fn is_nm_ip_route_removed(
-    for cur_route in cur_routes {
-        let mut cur_route_norm = cur_route.clone();
-        if cur_route_norm.table.is_none() {
-            cur_route_norm.table = Some(DEFAULT_TABLE_ID);
```
I added this workaround to prevent the full activation on the existing connection, https://github.com/nmstate/nmstate/pull/2639, but I think that your approach is better, because you convert the predefined default values to None at creation time, which will guarantee that the route table, next-hop will have the consistent default value for routes, therefore, the comparisons are always successful.

cathay4t commented on rust/src/lib/nm/nm_dbus/connection/route.rs:
```diff
@@ -26,20 +33,32 @@ pub struct NmIpRoute {
+        let empty_next_hop_val =
+            match dest.as_deref().unwrap_or("").parse::<std::net::IpAddr>() {
+                Ok(std::net::IpAddr::V4(_)) => IPV4_EMPTY_NEXT_HOP_ADDR,
+                Ok(std::net::IpAddr::V6(_)) => IPV6_EMPTY_NEXT_HOP_ADDR,
```
The `nm_dbus` crate should only provide DBUS abstraction without any smart modifications, these smart modification should be done at `nm` level.

```patch
diff --git a/rust/src/lib/nm/query_apply/route.rs b/rust/src/lib/nm/query_apply/route.rs
index d1d6be15..41962016 100644
--- a/rust/src/lib/nm/query_apply/route.rs
+++ b/rust/src/lib/nm/query_apply/route.rs
@@ -3,6 +3,8 @@
 use super::super::nm_dbus::{NmConnection, NmIpRoute};
 
 const DEFAULT_TABLE_ID: u32 = 254; // main route table ID
+const EMPTY_NEXT_HOP_ADDRV4: &str = "0.0.0.0";
+const EMPTY_NEXT_HOP_ADDRV6: &str = "::";
 
 pub(crate) fn is_route_removed(
     new_nm_conn: &NmConnection,
@@ -33,16 +35,39 @@ pub(crate) fn is_route_removed(
     )
 }
 
+fn sanitize_nm_ip_route(route: &mut NmIpRoute) {
+    if route.table.is_none() {
+        route.table = Some(DEFAULT_TABLE_ID);
+    }
+    if route.next_hop.as_deref() == Some(EMPTY_NEXT_HOP_ADDRV4)
+        || route.next_hop.as_deref() == Some(EMPTY_NEXT_HOP_ADDRV6)
+    {
+        route.next_hop = None;
+    }
+}
+
 fn is_nm_ip_route_removed(
     new_routes: &[NmIpRoute],
     cur_routes: &[NmIpRoute],
 ) -> bool {
+    let new_routes: Vec<NmIpRoute> = new_routes
+        .iter()
+        .map(|r| {
+            let mut route = r.clone();
+            sanitize_nm_ip_route(&mut route);
+            route
+        })
+        .collect();
+    let cur_routes: Vec<NmIpRoute> = cur_routes
+        .iter()
+        .map(|r| {
+            let mut route = r.clone();
+            sanitize_nm_ip_route(&mut route);
+            route
+        })
+        .collect();
     for cur_route in cur_routes {
-        let mut cur_route_norm = cur_route.clone();
-        if cur_route_norm.table.is_none() {
-            cur_route_norm.table = Some(DEFAULT_TABLE_ID);
-        }
-        if !new_routes.contains(&cur_route_norm) {
+        if !new_routes.contains(&cur_route) {
             return true;
         }
     }
```

cathay4t reviewed (changes requested):
The CI test case is still NM specific but the original issue did not use any nmcli command but pure nmstate YAML.
The key reproducer is a route next hop address pointing to a IP not in the CIDR of interface's IP, which NM will additional route with dst 0.0.0.0.

Hence I think the integration test case should be a minimized nmstate only two YAML apply, instead of this nmcli hacks.

ihuguet committed:
route: correctly compare the route's table and metric

When the NM route's table is undefined, NM will use the value defined in
ipv4/6.route-table, so we need to use that value for the normalization.

When the route's metric is undefined, NM will use the value defined in
ipv4/6.route-metric, so use it too.

If these route-table and route-metric properties are set to "default",
the final used value depends on global config and even on the device
type. We cannot know what value will be used in that case, so fail the
comparison so the routes are removed and added again with table and
metric set.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

ihuguet committed:
route: correctly compare the route's next-hop

In nmstate routes without next-hop are represented with "0.0.0.0" or
"::". This is because we need to differentiate the case "I don't care
about the next-hop for routes matching" (next-hop=None) and "This route
doesn't have next-hop" (next-hop=0.0.0.0).

However, NM doesn't do the same because it doesn't need that
differentiation. If a NM profile contains a direct route, without
next-hop, it will just leave it empty, thus we get next-hop=None.

This causes that comparing an NmIpRoute created from a nmstate route
with an NmIpRoute created from an existing NM profile will fail because
of this difference. Fix it by converting "0.0.0.0" / "::" to None when
creating a NmIpRoute from an nmstate route. This is the right
representation for it at `nm` layer, anyway.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

ihuguet committed:
nm: enforce setting table and metric on routes

Nmstate's documentation about static routes' table-id says that if a
value is not defined the main table (254) is used. However, this was not
actually enforced because it the NMConnection defines the
ipv4/6.route-table, then the main table is not used. Even if it's not
defined, NM might use a different table if that is chosen in its global
configs.

Moreover, nmstate sometimes define ipv4/6.route-table, but only for
dynamically obtained routes, if "auto-route-table-id" is defined.

Enforce table=254 for all static routes that doesn't define a
"table-id", ensuring that they are never affected by
"auto-route-table-id".

Enforce metric=1024 for ipv6, as this is the value that will be coherced
by kernel if we set metric=0. This fixes is_route_removed failing
because it expected metric=0 and gets metric=1024.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

cathay4t commented on tests/integration/conftest.py:
```diff
@@ -162,6 +163,12 @@ def eth1_up(test_env_setup):
 
 
+@pytest.fixture(scope="function")
+def dummy0_up(test_env_setup):
```
Please store this fixture to `testlib/dummy.py` or `route_test.py`. This `conftest.py` is for fixture used by all test cases.
